mvc/models/model-binding.rst

```diff
@@ -1,8 +1,113 @@
+
+.. image:: model-binding/_static/routevalues-in-watch-window.png
+
+Since the  routing pattern looks like this, ``{controller=Home}/{action=Index}/{id?}``, the route ``movies/edit/2`` maps to the ``Movies`` controller, and its ``Edit`` action method. It also accepts an optional ``id`` of 2 as the image shows here:
+
+.. image:: model-binding/_static/model-binding.png
+
+This is due to MVC's use of conventions. So the code for the action method should look something like this: 
+
+.. literalinclude:: model-binding/sample/src/MVCMovie/Controllers/MoviesController.cs
+   :language: c#
+   :lines: 63-76  
+   
+.. note:: The strings in the URL are not case sensitive.
+
+Routing knows which action method to use, so it checks for model binding attributes applied to the action method's arguments. If it finds a model binding attribute it will use the data source specified by the attribute. Otherwise, for each argument (a primitive type for now, more on complex types later), it looks in its data sources for something with a matching name. In the above example the only argument is an integer named `id`, so that is what the binder will look for. Below is a list of the data sources in the order that the model binder looks through them:
+ 
+#. ``Request.Form``: These are form values that go in the HTTP request using the POST method.
+#. ``RouteData.Values``: Contains information about each segment of the route. Route segments are separated by a forward slash ``/``.
+#. ``Request.QueryStrings``: Data to the right of a question mark (?) in a URL, separated by ampersands (&). For example, genre and subgenre are query strings: http://contoso.com/movies?genre=Drama&subgenre=Crime
+
+.. note:: Form, RouteData, and QueryStrings are all stored as name-value pairs.
+
+In the example, the model binder is loos through the Request.Form first for a key named `id`. In this case it didn't find anything, so it moved on to RouteData.Values looking for a key with a name of `id`. In our example, it's a match, since the last portion of the route has a key value of `id`, and it is an integer. Binding happens, and the value of the id parameter becomes 2. If the action method's parameter were a complex type such as the ``Movie`` type, which likely contains both primitive and complex types, MVC's model binder will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types through its tree, and comparing any primitive members' names against the names of keys in data sources. The model binder can bind objects that contain nested objects for multiple levels. 
+
+.. note:: When the binder finds the first matching name in the data source, it stops looking for more elements matching that name, and moves onto the next property or item in the parameter list. 
+ 
+.. note:: If for some reason binding can't happen, it quietly fails.
```

Comment:
This doesn't sound quite right. Generally model binding issues are recorded as ModelState errors. What kinds of errors are you referring to?

Maybe a better statement would be that model binding makes a best effort to bind as much as it can. I.e. it tries not to fail fast.